File: leetcode/2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.java
class Solution {
    public String shortestBeautifulSubstring(String s, int k) {
        int stringLength = s.length();
        String result = "";
      
        // Iterate through all possible starting positions
        for (int startIndex = 0; startIndex < stringLength; ++startIndex) {
            // Try all possible ending positions (minimum length k to contain k ones)
            for (int endIndex = startIndex + k; endIndex <= stringLength; ++endIndex) {
                // Extract the current substring
                String currentSubstring = s.substring(startIndex, endIndex);
              
                // Count the number of '1's in the current substring
                int onesCount = 0;
                for (char character : currentSubstring.toCharArray()) {
                    onesCount += character - '0';  // Convert '1' to 1, '0' to 0
                }
              
                // Check if this substring has exactly k ones and is better than current result
                if (onesCount == k && 
                    (result.isEmpty() ||                                    // First valid substring found
                     endIndex - startIndex < result.length() ||             // Shorter than current result
                     (endIndex - startIndex == result.length() &&          // Same length but
                      currentSubstring.compareTo(result) < 0))) {           // lexicographically smaller
                    result = currentSubstring;
                }
            }
        }
      
        return result;
    }
}

File: leetcode/2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.js
/**
 * Finds the shortest beautiful substring containing exactly k '1's.
 * If multiple substrings have the same length, returns the lexicographically smallest one.
 *
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
function shortestBeautifulSubstring(s, k) {
    const stringLength = s.length;
    let shortestSubstring = '';

    // 모든 시작 위치 탐색
    for (let startIndex = 0; startIndex < stringLength; startIndex++) {

        // 최소 길이가 k인 부분 문자열부터 탐색
        for (
            let endIndex = startIndex + k;
            endIndex <= stringLength;
            endIndex++
        ) {
            // 현재 부분 문자열
            const currentSubstring = s.slice(startIndex, endIndex);

            // 현재 부분 문자열의 1 개수 계산
            const onesCount = currentSubstring
                .split('')
                .filter(char => char === '1')
                .length;

            // 조건을 만족하는지 확인
            const isValidSubstring = onesCount === k;

            const isFirstValid = shortestSubstring === '';

            const isShorter =
                endIndex - startIndex < shortestSubstring.length;

            const isSameLengthButSmaller =
                endIndex - startIndex === shortestSubstring.length &&
                currentSubstring < shortestSubstring;

            // 정답 갱신
            if (
                isValidSubstring &&
                (isFirstValid || isShorter || isSameLengthButSmaller)
            ) {
                shortestSubstring = currentSubstring;
            }
        }
    }

    return shortestSubstring;
}